### ISSUE TITLE
Modular packages

### ISSUE THREAD

**Issue by jaredpalmer** [algolia/react-instantsearch#640]
**Feature: What is your use case for such a feature?**
Keeping my bundle small AF

**Feature: What is your proposed API entry? The new option to add? What is the behavior?**

Refactor widgets and core so that individual widgets can be added a la carte. Resulting package structure might look like:

- packages
  - react-instantsearch: Shell package that requires and exports core, utils, and all widgets
  - react-instantsearch-core: Just the guts, no UI or css
  - react-instantsearch-utils: shared utils across packages
  - react-instantsearch-widgets: all widgets
  - react-instantsearch-searchbox: Just the `<SearchBox />`, peer dependency on core.
  - react-instantsearch-menu: Just the `<Menu />`, peer dependency on core.

Change would non-breaking, require no modifications to the current docs, and keep install counts on NPM in tact. However, it would allow devs to add widgets on an a la carte basis and minimize their bundle size.

**Comment by Haroenv** [comment]
Did you try the /es/ path? It was made with the assumption bundles could treeshake. But as you say, it might be more efficient to have separate modules

**Comment by samouss** [comment]
Like @Haroenv said you can use the `/es` import (like below). Bundler who support tree-shaking will be able to remove all unused import from the final bundle.

```js
import { InstantSearch, SearchBox } from 'react-instantsearch/es/dom';
```

In fact it's already plan to split the package into multiple one. The split that we thought  is a bit diffent though:

- `react-instantsearch`: will be the core features 
- `react-instantsearch-connectors`: will be the connectors
- `react-instantsearch-dom`: will be the DOM widgets
- `react-instantsearch-native`: will be the native widgets
- `react-instantsearch-server`: will be the server specific

The advantage of this will be to enable import of ES modules by default for bundler users, without specifying the `/es` path.

**Comment by davidfurlong** [comment]
Is the `/es` import route documented somewhere, I couldn't find it when looking for how to reduce the react-instantsearch size in my bundle?

**Comment by Haroenv** [comment]
@davidfurlong, both algoliasearch and algoliasearch-helper should be in the bundle, since they're used by the library. I'd gladly take a look at a reproduced example to see if something weird could have happened with your specific webpack setup though

**Comment by samouss** [comment]
Indeed @Haroenv is right, there is an ongoing work to split the packages into multiple ones and reduce the size of the bundle. With the current status the lib is not correctly tree shaken. It will be fixed during the coming months but it will only work for Webpack 4+ (we will take leverage of the [`sideEffects`](https://webpack.js.org/guides/tree-shaking/#mark-the-file-as-side-effect-free) attribute).

**Comment by davidfurlong** [comment]
Ok awesome thanks for the update!

**Comment by samouss** [comment]
We released a new version `5.2.0` that split the package into multiple ones. Those new packages are compatible with tree-shaking (on Webpack 4+). You can find more information about the migration process & motivations [in the documentation](https://community.algolia.com/react-instantsearch/guide/Migration_guide_package_import.html). Don't hesitate to give us feedback!
